api/src/main/java/org/apache/iceberg/expressions/ResolvedReference.java

```diff
@@ -0,0 +1,78 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.expressions;
+
+import org.apache.iceberg.Schema;
+import org.apache.iceberg.exceptions.ValidationException;
+import org.apache.iceberg.types.Types;
+
+public class ResolvedReference<T> extends NamedReference<T> {
+  private final int fieldId;
+
+  public ResolvedReference(String name, int fieldId) {
+    super(name);
+    this.fieldId = fieldId;
+  }
+
+  public int fieldId() {
+    return fieldId;
+  }
+
+  @Override
+  public BoundReference<T> bind(Types.StructType struct, boolean caseSensitive) {
+    Schema schema = struct.asSchema();
+    Types.NestedField field = schema.findField(fieldId);
+    ValidationException.check(
+        field != null,
+        "Cannot find field with id '%s' in struct: %s, since we are resolving based on ID",
```

Comment:
This doesn't quite follow the project guidelines for clarity. One issue is that who "we" refers to is unclear and probably distracting. Instead prefer direct statements, like "Cannot find field by id %s in struct: %s". Also, because this is a `ResolvedReference` that is being bound, I don't think that it is necessary to separately state that resolution is based on ID. That should be clear from the first part, "Cannot find field with/by id".